react-components/src/architecture/concrete/primitives/box/BoxDragger.ts

```diff
@@ -241,52 +220,85 @@ export class BoxDragger extends BaseDragger {
-    size.copy(this._sizeOfBox);
-    center.copy(this._centerOfBox);
+    const originalBox = this._originalBox;
+    const { box } = this._domainObject;
+    box.copy(originalBox);
+    const { size, center } = box;
 
     // Apply the change
     size.add(deltaSize);
-    this._domainObject.forceMinSize();
+    box.forceMinSize();
 
-    if (size.lengthSq() === this._sizeOfBox.lengthSq()) {
+    if (size.lengthSq() === originalBox.size.lengthSq()) {
       return false; // Nothing has changed
     }
     // The center of the box should be moved by half of the delta size and take the rotation into account.
-    const newDeltaSize = newVector3().subVectors(size, this._sizeOfBox);
+    const newDeltaSize = newVector3().subVectors(size, originalBox.size);
     const deltaCenter = newDeltaSize.divideScalar(2);
     deltaCenter.multiply(this._cornerSign);
-    deltaCenter.applyMatrix4(rotationMatrix);
+    deltaCenter.applyMatrix4(this._rotationMatrix);
     center.add(deltaCenter);
     return true;
   }
 
   private rotate(ray: Ray, shift: boolean): boolean {
-    const endPoint = ray.intersectPlane(this._planeOfBox, newVector3());
+    const domainObject = this._domainObject;
+    if (!domainObject.canRotateComponent(this._face.index)) {
+      return false;
+    }
+    const endPoint = ray.intersectPlane(this._planeOfFace, newVector3());
     if (endPoint === null) {
       return false;
     }
-    const center = this._planeOfBox.projectPoint(this._centerOfBox, newVector3());
-    const centerToStartPoint = newVector3().subVectors(this.point, center);
-    const centerToEndPoint = newVector3().subVectors(endPoint, center);
+    const centerToStartPoint = newVector3().subVectors(this.point, this._centerOfFace).normalize();
+    const centerToEndPoint = newVector3().subVectors(endPoint, this._centerOfFace).normalize();
+    const cross = newVector3().crossVectors(centerToEndPoint, centerToStartPoint).normalize();
 
-    // Ignore Z-value since we are only interested in the rotation around the Z-axis
-    const deltaAngle = horizontalAngle(centerToEndPoint) - horizontalAngle(centerToStartPoint);
+    let deltaAngle = centerToEndPoint.angleTo(centerToStartPoint);
+    if (this._normal.dot(cross) > 0) {
+      deltaAngle = -deltaAngle;
+    }
+    const originalBox = this._originalBox;
+    const { box } = domainObject;
+    box.copy(originalBox);
 
-    // Rotate
-    let zRotation = forceBetween0AndPi(deltaAngle + this._zRotationOfBox);
     if (shift) {
-      let degrees = radToDeg(zRotation);
-      degrees = round(degrees, CONSTRAINED_ANGLE_INCREMENT);
-      zRotation = degToRad(degrees);
+      let oldAngle;
+      if (this._face.index === 0) {
+        oldAngle = originalBox.rotation.x;
+      } else if (this._face.index === 1) {
+        oldAngle = originalBox.rotation.y;
+      } else if (this._face.index === 2) {
+        oldAngle = originalBox.rotation.z;
+      } else {
+        return false;
+      }
```

Comment:
I must admit that I don't understand what the `shift` boolean does here :sweat_smile: I assume it says whether the shift button is pressed or not, but I don't understand its effect. Could this part be put in a separate named function to make it more clear what's happening?

Reply:
Renamed to isShiftPressed. Also made everything clearer.